### PR TITLE
Refactor bevy_mikktspace into idiomatic Rust

### PR DESCRIPTION
**Pull request by DJMcNab**
# Objective

- `bevy_mikktspace`'s use of unsafe is concerning.
- Eventually, we want this to be `#[forbid(unsafe_code)]`
- The refactor needs to maintain the same behaviour.

## Solution

- [x] Do small refactoring on the unsafe version (e.g. moving variables down to their declarations, adding comments)
- [x] Chip away at the unsafe code
- [x] Refactor some of the gnarlier parts to use standard facilities

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Yes please. Fundamentally from what I can see there's nothing in this algorithm that should require the use of `unsafe`.

**Comment by alice-i-cecile** [comment]
I'm giving this the `Unsound` label. I haven't *proven* that there's UB in the original auto-translated code, I would be very very surprised if there was none. Even if there isn't, the safety hazards of having this code around in the current form are very high due to the extreme challenges working with the existing code correctly.

**Comment by DJMcNab** [comment]
bors try

Failure seems to be spurious?

**Comment by superdump** [comment]
I would like to review this before it is merged. Just mentioning that so it doesn’t get merged before I’ve taken a look. I’m concerned about the numeric behaviour of any changes. I know @DJMcNab is aware of this concern and I trust that care has been taken. Still I feel it is easy to unintentionally change arithmetic behaviour. Ideally there would be a comprehensive set of regression tests for this kind of thing, as @DJMcNab suggests, comparing against the behaviour of the original the scope of verifying numeric results are the same. I suppose as we have the flight helmet model in the repo, one thing we could do is to generate tangents for that using the original C code and store those in the repo, and generate tangents using this code and compare. It wouldn’t be a set of tests targeted at edge/corner cases but it would at least test something.

> Note to reviewers: This isn't used at all internally, so hopefully there are no behaviour changes, but I don't care significantly if there are :).

I don’t understand this comment. What isn’t used internally?

**Comment by DJMcNab** [comment]
So what I mean is my priority is removing the unsound implemenation more than having exactly the same behaviour as the C code.

As an examplel, the original C code is completely careless in their handling of `NaN`/`Infinity`, and so I'm fairly sure we have different behaviour there (more panicky). However, I think the original's behaviour would still be completely wrong anyway.
As far as I know, the only changes I've made to the behaviour are cases where `NaN`/`Infinity` (/maybe subnormals) crop up. The vast majority of the code doesn't do any floating point handling, and is just moving data around. The parts that do floating point operations, I think I've done an exact translation.

**Comment by superdump** [comment]
> So what I mean is my priority is removing the unsound implemenation more than having exactly the same behaviour as the C code.

How do you define behaviour? It is only the numeric results that I care about being consistent. How it gets there, aside from performance, code maintenance, API ergonomics, etc, is secondary.

> As an examplel, the original C code is completely careless in their handling of `NaN`/`Infinity`, and so I'm fairly sure we have different behaviour there (more panicky). However, I think the original's behaviour would still be completely wrong anyway. As far as I know, the only changes I've made to the behaviour are cases where `NaN`/`Infinity` (/maybe subnormals) crop up.

NaN/infinity aren’t always invalid in shader land. There is a trick used to discard vertices by setting their position to NaN for example. I’d be careful about deciding to explicitly check and handle NaNs unless it is very clear that a NaN at that point is invalid and has to be handled.

> The vast majority of the code doesn't do any floating point handling, and is just moving data around in hilariously evil ways. The parts that do floating point operations, I think I've done an exact translation.

Right, totally down with making how the data is moved around be safe and rusty. And if indeed the float operations are an exact translation (including allowing NaNs or whatever, I guess, until proven that handling them differently is the correct thing to do) then this sounds great. Just saying this for the purpose of alignment. :)

**Comment by DJMcNab** [comment]
As an example, as far as I can tell, if there's a single vertex with infinite position, [this line](https://github.com/mmikk/MikkTSpace/blob/3e895b49d05ea07e4c2133156cfa94369e19e409/mikktspace.c#L604) means that none of the vertices get merged.
However, identical vertices being merged is relied upon for correctness. 

So, assuming that all positions are finite, what next?

Following through the code, eventually you get to calling [`m_setTSpace`](https://github.com/mmikk/MikkTSpace/blob/3e895b49d05ea07e4c2133156cfa94369e19e409/mikktspace.h#L102-L103) with `NaN` or `Infinity` vectors, which breaks the [contract of that method](https://github.com/mmikk/MikkTSpace/blob/master/mikktspace.h#L92).

This is in C api rules, which best as I can tell is that any input which gets garbage out must be an illegal input.

The only mention of NaN handling in mikktspace I can find is [here](https://lists.blender.org/pipermail/bf-blender-cvs/2016-September/090385.html), which prevents an infinite loop when they occur. In the original paper, the word `NaN` does not appear, not does `Infinity` except in a single mention for (as best as I can tell) something unrelated. 

Additionally, as best as I can tell from a quick check, weird values are [forbidden](https://www.khronos.org/registry/glTF/specs/2.0/glTF-2.0.html#accessor-data-types) in GLTF.

Basically, I'm pretty confident that breaking behaviour in the prescense of NaN/infinity is fine.

**Comment by DJMcNab** [comment]
I'm not likely to drive this forward, so I'll defer to https://github.com/bevyengine/bevy/pull/9050
